api/src/main/java/org/pmiops/workbench/actionaudit/auditors/UserServiceAuditAdapterImpl.java

```diff
@@ -39,13 +42,29 @@ public UserServiceAuditAdapterImpl(
 
   @Override
   public void fireUpdateDataAccessAction(
-      DbUser targetUser, DataAccessLevel dataAccessLevel, DataAccessLevel previousDataAccessLevel) {
+      DbUser targetUser,
+      DataAccessLevel dataAccessLevel,
+      DataAccessLevel previousDataAccessLevel,
+      AgentType agentType) {
+    // In the event of an offline cronjob, the request may have been made anonymously and there is
+    // no user identity. This should only happen for system agent driven audit events.
+    Optional<DbUser> agent = Optional.ofNullable(dbUserProvider.get());
+    if (agent.isPresent() && AgentType.SYSTEM == agentType) {
+      log.warning(
+          String.format(
+              "found unexpected request scope user '%s' for SYSTEM agent audit action",
+              agent.get().getUsername()));
+    } else if (!agent.isPresent() && AgentType.SYSTEM != agentType) {
+      log.warning(
+          String.format("missing request scope user for audit action agent type of %s", agentType));
+    }
+
     actionAuditService.send(
         ActionAuditEvent.builder()
             .timestamp(clock.millis())
-            .agentType(AgentType.ADMINISTRATOR)
-            .agentId(dbUserProvider.get().getUserId())
-            .agentEmailMaybe(dbUserProvider.get().getUsername())
+            .agentType(agentType)
+            .agentId(agent.map(DbUser::getUserId).orElse(0L))
+            .agentEmailMaybe(agent.map(DbUser::getUsername).orElse(null))
```

Comment:
nit: typically the pattern I use with optionals adn builders is 
`agent.ifPresent(eventBuilder::agentEmailMaybe)`. Downside is you have to declare the builder.

Reply:
That's neat, didn't know about that method. In this case I think I prefer how it's currently written though.